packages/modal/Modal.tsx

```diff
@@ -14,62 +15,113 @@ export class Modal extends Component<ModalProps> {
-    this.isOpen = !this.isOpen;
+    this.isOpen = false;
     emit(this,'modalClose')
   }
+  private focusModal() {
+    this.modalElement.focus();
+  }
+
+  private handleDocumentFocus(event: FocusEvent) {
+    if (this.modalElement && !this.modalElement.contains(event.target as Node)) {
+      event.stopImmediatePropagation();
+      this.focusModal();
+    }
+  }
+
+  private preventModalBlur() {
+    document.addEventListener("focus", this.handleDocumentFocus.bind(this), true);
+  }
+  private allowModalBlur() {
+    document.removeEventListener("focus", this.handleDocumentFocus.bind(this), true);
+  }
+
   connectedCallback(){
     super.connectedCallback();
     this.handleEsc = this.handleEsc.bind(this);
     this.handleModalClose = this.handleModalClose.bind(this);
+    this.focusModal = this.focusModal.bind(this);
   }
   renderCallback() {
-    const {isOpen} = this;
+    const {isOpen, closeTitle} = this;
     return [
       <style>{styles}</style>,
-      isOpen && <div class="c-overlay c-overlay--fullpage"/>,
+      isOpen &&
+      <div class="c-overlay c-overlay--fullpage"
+           tabIndex={-1}
+           onFocus={this.focusModal} />,
       isOpen &&
       <div ref={(_ref: HTMLDivElement)=>this.modalElement=_ref}
-           tabIndex={0}
+           tabIndex={-1}
            class="o-modal"
+           role="dialog"
+           aria-labelledby="modal-heading"
+           aria-describedby="modal-body"
            onKeydown={this.handleEsc}
       >
-       <Card>
-         <Button
-           slot="dismiss"
-           onClick={this.handleModalClose}>
-           x
-         </Button>
-         <div slot="title">
-           <slot name="title"></slot>
-         </div>
-         <div slot="body">
-           <slot></slot>
-         </div>
-         <div slot="footer">
-           <slot name="footer"></slot>
-         </div>
-       </Card>
+        <Card>
+          <Button
+            slot="dismiss"
+            aria-label={closeTitle}
+            type="close"
+            onClick={this.handleModalClose}>
+            x
+          </Button>
+          <div slot="heading" id="modal-heading">
+            <slot name="title"></slot>
+          </div>
+          <div slot="body" id="modal-body">
+            <slot></slot>
+          </div>
+          <div slot="footer">
+            <slot name="footer"></slot>
+          </div>
+        </Card>
       </div>
     ]
   }
 
   renderedCallback() {
     if ( this.isOpen ) {
-      this.modalElement.focus();
+      this.lastActiveElement = this.deepActiveElement();
+
+      this.focusModal();
+
+      this.preventModalBlur();
+    } else {
+      this.allowModalBlur();
+
+      if (this.lastActiveElement) {
+        this.lastActiveElement.focus();
+      }
     }
   }
+
+  private deepActiveElement(): HTMLElement {
+    let activeElement = document.activeElement;
+    while (activeElement && activeElement.shadowRoot && activeElement.shadowRoot.activeElement) {
```

Comment:
You just need to find the active element (the one that is **actually** focused).

Otherwise first WebComponent element will be taken and returned and you would have to make sure, to handle focus of that element to focus what ever needs to be focused.

And there might be multiple focus-able items within the component, so you might not be able to know which is correct.